client-misk/src/main/kotlin/app/cash/backfila/client/misk/internal/HibernateBackfillOperator.kt

```diff
@@ -0,0 +1,354 @@
+          }!!
+          matchingCount = result[0] as Long?
+          end = boundingMax!!
+        } else {
+          // We got an id, so there's exactly batchSize results.
+          matchingCount = batchSize
+          end = batchEndPkey
+        }
+
+        // Get start pkey and scanned record count for this batch.
+        val result = queryFactory.dynamicQuery(backfill.entityClass).apply {
+          addBoundingMin(this)
+          dynamicAddConstraint(pkeyProperty, LE, end)
+        }.dynamicUniqueResult(session) { criteriaBuilder, queryRoot ->
+          criteriaBuilder.tuple(
+              criteriaBuilder.min(backfill.getPrimaryKeyPath(queryRoot)),
+              criteriaBuilder.count(queryRoot))
+        }!!
+        val start = result[0].toString()
+        val scannedCount = result[1] as Long
+
+        TxResult(end,
+            Batch.Builder()
+                .batch_range(KeyRange.Builder()
+                    .start(start.encodeUtf8())
+                    .end(end.toString().encodeUtf8())
+                    .build())
+                .scanned_record_count(scannedCount)
+                .matching_record_count(matchingCount ?: 0)
+                .build())
+      }
+      if (txResult.end == boundingMax) {
+        // Reached the end of this bounding range, null it out so a new one is computed when more
+        // batches are requested.
+        boundingMax = null
+      }
+      previousEndKey = txResult.batch.batch_range.end
+      return txResult.batch
+    }
+  }
+
+  override fun runBatch(request: RunBatchRequest): RunBatchResponse {
+    val config = parametersOperator.constructBackfillConfig(
+        request.parameters, request.dry_run)
+
+    val pkeys = partitionProvider.transaction(request.partition_name) { session ->
+      val minId = Id<E>(request.batch_range.start.utf8().toLong())
+      val maxId = Id<E>(request.batch_range.end.utf8().toLong())
+      val pkeyProperty = backfill.primaryKeyHibernateName()
+      val idList = backfill.backfillCriteria(config).apply {
+        dynamicAddConstraint(pkeyProperty, GE, minId)
+        dynamicAddConstraint(pkeyProperty, LE, maxId)
+        dynamicAddOrder(pkeyProperty, true)
+      }.dynamicList(session, listOf(pkeyProperty))
+
+      // Convert them back to a list of Pkey
+      @Suppress("UNCHECKED_CAST") // Return type from the query should always match.
+      idList.map { it.single() } as List<Pkey>
+    }
+    backfill.runBatch(pkeys, config)
+
+    return RunBatchResponse.Builder().build()
+  }
+
+  private fun Backfill<*, *, *>.getPrimaryKeyPath(queryRoot: Root<*>): Path<Number> {
+    val fields = primaryKeyHibernateName().split('.')
+    var path = queryRoot as Path<Number>
+    for (field in fields) {
+      path = path.get(field)
+    }
+    return path
+  }
+
+  @Singleton
+  class HibernateBackend @Inject constructor(
+    private val injector: Injector,
+    @ForBackfila private val backfills: MutableMap<String, KClass<out Backfill<*, *, *>>>,
```

Comment:
I assume you won't be binding dynamo backfills in this map

Reply:
Yep. Separate map.